### ISSUE TITLE
[Proposal] Call for feedback on new button name db.

### ISSUE THREAD

**Issue by p-groarke** [gabomdq/SDL_GameControllerDB#178]
I am working on a new (separate) db file to map buttons to appropriate names. This will allow you to display appropriate graphics for controllers. The goal isn't to be perfect, but to at least show the right name for buttons that affect gameplay.

Here is a draft of the format. I am looking for feedback on the available button names (am I missing a controller?), missing hints, or any unforeseen issues that would arise from this format.

```
# Game Controller Button Name DB
# Source: https://github.com/gabomdq/SDL_GameControllerDB
#
# Help
# Format : guid,name,a,b,hint,leftshoulder,lefttrigger,rightshoulder,righttrigger,x,y,platform
#
# Button strings (a,b,x,y) are :
#     A, B, C, T, U, X, Y, Z, CROSS, CIRCLE, SQUARE, TRIANGLE, 1, 2, 3, 4, 5, 6
#
# Shoulder strings (leftshoulder,rightshoulder) are :
#     LB, L1, RB, R1, L, R, 5, 6, 7, 8
#
# Trigger strings (lefttrigger,righttrigger) are :
#     LT, L2, RT, R2, ZL, ZR, Z, 5, 6, 7, 8, 9, 10
#
# Hints are :
#     XBOX, PS, NINTENDO, SEGA, ARCADE
#
# Example
# 03000000022000000090000000000000,8Bitdo NES30 Pro,a:B,b:A,hint:NINTENDO,leftshoulder:L1,lefttrigger:L2,rightshoulder:R1,righttrigger:R2,x:Y,y:X,platform:Windows,
```

Again, the goal is not to be perfect, but to at least show the right names for in-game buttons (start,home,select aren't included). The format should mimic the existing db as closely as possible.

**Comment by eddieray** [comment]
Great idea!

**Comment by p-groarke** [comment]
Good point on the pedals and shifter boxes! I'll have to look into that and see if it can fit in with the current setup. Thx.

The underside or extra buttons, since they don't actually map to the xbox controller, are ignored. At that point, you'd want to detect the controller directly in your game and do fancy things. The same can be said with gyros, accelerometers or touchpads. The db is for basic (but full) support. The extras aren't supported.

The hint is indeed for graphics. Here is a pack of awesome graphics for many different controllers. It's released with cc0 license (public domain, fully free) https://opengameart.org/content/free-keyboard-and-controllers-prompts-pack

In the extreme long-term (aka years), I'd like to provide some generic svg graphics for all buttons, but not styled to the brands. If that makes any sense. For now though, graphics are out of the question ;)

**Comment by p-groarke** [comment]
@eddieray Quick note on driving peripherals, they are considered joysticks and don't use this database (which is for gamepads only). So we are all good there.

**Comment by mbenkmann** [comment]
I am working on generic SVG for controllers for a game I'm creating. The idea is to do away with names altogether and instead use parts of the whole controller image with the respective button(s) highlighted. So after the user has selected (or it has been pulled from a database) the proper image for his controller, all references to buttons are purely via small graphics.

With respect to your new database, I don't think button name mappings are enough. I would definitely add controller geometry to the data. Not precise geometry but simply a list of elements that are present on the controller. Look at this image

![foo](http://www.winterdrache.de/data/g4237.png)

It reasonably covers many common controllers, when you combine it with a list of elements that are actually present. So for each controller you have a list of elements from all the possible elements together with the name.

**Comment by mbenkmann** [comment]
This approach would allow game programmers to auto-generate controller images with their own graphics assets. So I could put that database in my game and it would tell me e.g. that the XBox 360 controller has no touchpad and only 4 main buttons in the top right. I would then paint my SVG with the non-existent elements removed.

**Comment by p-groarke** [comment]
You have to remember the purpose of this database is to map everything to a 360 controller. That is the geometry. There wont be any funky controls information present (like the DS4 touchpad). However, you can get this information easily with controller names in the database if you want.

Generating your own graphics is interesting, but a marginal use case nonetheless. The hints are there to help with what you want to do though.

If you read a button name mapping that has hint SEGA for example, you know where the buttons are going to be (somewhat). The image below is the current guideline. However that is new, so not all mappings will be correct (thats another problem that will need dealing with).

Basically, you should be able to get most of the information you need from what is present, but it will not be perfect. Realistically, it cannot. There are too many controllers out there with various geometries. Asking contributors to measure their controller is also too much to ask.

![mappings](https://github.com/gabomdq/SDL_GameControllerDB/raw/master/data/mapping_guide.png)

**Comment by cxong** [comment]
Has this idea been abandoned? It overlaps with this project: https://github.com/cxong/SDL_JoystickButtonNames which aims to map button names as well as button colours.

**Comment by p-groarke** [comment]
@cxong Pretty much yes. The problem is there was no standard layout recommended in the database till recently, and we have no way of knowing what buttons users mapped to the xbox buttons. You can check branch `button` for some layouts I had created. I personally wont be taking care of this.

The project you linked to has been inactive for ~3 years, and it is also quite different than my original goal.

**Comment by JayFoxRox** [comment]
**The mapping recommendation for SEGA and Arcade looks flawed.**

The gamecontroller mimics the Xbox 360 gamepad. The original Xbox is similar to SEGA layout (6 face buttons).

Microsoft, the maker of the original Xbox and Xbox 360 layouts, made the official mapping between Xbox and Xbox 360. Due to similarity with the SEGA pad, we have an implicit, official mapping between Xbox 360 and SEGA.

---

Where SEGA has Z / C, original Xbox has white / black. In original Xbox games on Xbox 360 (and also many successors to existing games), these games use LB=White and RB=Black to replace the missing face buttons.

This would mean:

* Xbox 360 &rarr; original Xbox: LB=White & RB=Black
* SEGA &harr; original Xbox: Z=White & C=Black

So, shouldn't it be:

* SEGA &rarr; Xbox 360: Z=White & C=Black

So SEGA would be:

```
X Y LB
A B RB
```

And a similar mapping for Arcade:

```
X Y LB LT
A B RB RT
```

That would also mean that all digital buttons will be mapped to digital buttons, leaving the analog triggers for other tasks.

**Comment by offalynne** [comment]
I agree with @JayFoxRox's proposal. Making the transition would prove tricky because in many cases it looks like maps do not match the existing spec and ie, follow the button labels on a given arcade stick's equivalent gamepad labels rather than the format proposed by the readme graphic.

Saturn pads have shoulder buttons, and it makes sense in my opinion to map LB/RB on Z/C and put the triggers on L and R.
